pandas/core/frame.py

```diff
@@ -988,6 +988,33 @@ def __dataframe_consortium_standard__(
         )
         return convert_to_standard_compliant_dataframe(self, api_version=api_version)
 
+    def __arrow_c_stream__(self, requested_schema=None):
+        """
+        Export the pandas DataFrame as an Arrow C stream PyCapsule.
+
+        This relies on pyarrow to convert the pandas DataFrame to the Arrow
+        format (and follows the default behaviour of ``pyarrow.Table.from_pandas``
+        in its handling of the index, i.e. store the index as a column except
+        for RangeIndex).
+        This conversion is not necessarily zero-copy.
+
+        Parameters
+        ----------
+        requested_schema : PyCapsule, default None
+            The schema to which the dataframe should be casted, passed as a
+            PyCapsule containing a C ArrowSchema representation of the
+            requested schema.
+
+        Returns
+        -------
+        PyCapsule
+        """
+        pa = import_optional_dependency("pyarrow", min_version="14.0.0")
+        if requested_schema is not None:
+            requested_schema = pa.Schema._import_from_c_capsule(requested_schema)
```

Comment:
Can't you use `pa.schema()` directly instead of `pa.Schema._import_from_c_capsule`?

Reply:
I don't think it will necessarily become public in the current form (but the

Reply:
I don't think it will necessarily become public in the current form (but the `_import_from_c` version has been used in many other external projects, so we won't just change those methods in pyarrow)

> Can't you use `pa.schema()` directly instead of `pa.Schema._import_from_c_capsule`?

Not directly, because we get a capsule here (we are inside the low-level dunder here), and `pa.schema()` doesn't accept capsules, only objects implementing `__arrow_c_schema__`. Of course we could have a small wrapper object that has the dunder method and returns the capsule, if we want to avoid using the `_import_from_c_capsule`. 

I brought this up in the past on the pyarrow side whether we need an "official" way to import capsules, in the last paragraph in https://github.com/apache/arrow/issues/38010, but we should maybe discuss that a bit more (or whether we just "bless" the `_import_from_c_capsule` as the official way to do this)